server/bound_queries.cc

```diff
@@ -163,22 +165,40 @@ struct UnorderedExecutor: public BoundSelectQuery::Executor {
                     return true;
                 };
 
-            if (allowMT) {
-                parallelMap(0, effectiveNumBucket, doBucket);
-            }
-            else {
-                for (int i = 0; i < effectiveNumBucket; ++i)
-                    doBucket(i);
-            }
+            parallelMap(0, effectiveNumBucket, doBucket);
         }
         else {
-            if (allowMT) {
-                parallelMap(0, rows.size(), doRow);
-            }
-            else {
-                for (int i = 0; i < rows.size(); ++i)
-                    doRow(i);
+            size_t upper =  rows.size();
+            if (limit != -1)
+                upper = std::min((size_t)(offset+limit), upper);
+
+            if (aggregateInParallel) {
+
+                parallelMap(offset, upper, doRow);
             }
+            else
+            {
+                //Todo: to reduce memory usage, we should fill blocks of output on worker threads
+                // in order as much as possible
+                // while calling the aggregator on the caller thread.
+                std::vector<std::tuple<NamedRowValue, std::vector<ExpressionValue> > > output(upper-offset);
```

Comment:
The validation that offset >= 0 and offset <= upper is probably done elsewhere but if it is not, I suggest we add an assert here before the allocation of upper - offset.